### PR TITLE
[91X] L1T Validation Module with Stage2 updates

### PR DESCRIPTION
**Pull request by ly615**
This PR includes the update of L1T collections in Run II  in order to fill the L1TriggerVsGen histograms.  The change is mostly done in Validation/L1T package, which is an important package for L1T RelVal. A corresponding L1T_postProcessor python script is created in Validation/Configuration to compute the L1T efficiencies w.r.t Gen objects.  
The PR has been tested on RelVal_Ztautau sample,  no problem shows up.

### CODE REVIEW

**Comment by perrotta** [comment]
@ly615 : could you please edit the title of this PR and write something more meaningful (i.e. explanation in plain English of what this PR is supposed to do)?

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
Validation stuff: code looks reasonable, the new validation plots show up in output

**Comment by rekovic** [comment]
@civanch, @mdhildreth, @dmitrijus, @vanbesien,
Can you please review and sign/comment.
We would like this to be merged this week.

**Comment by dmitrijus** [comment]
+1

**Comment by civanch** [comment]
+1

File: Validation/L1T/plugins/L1Validator.cc [hunk 1]
```diff
@@ -82,78 +81,141 @@ void L1Validator::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetu
...
-  _Hists.FillNumber(L1ValidatorHists::Type::TauJet, Taus->size());
-  _Hists.FillNumber(L1ValidatorHists::Type::Muon, Muons->size());
+  _Hists.FillNumber(L1ValidatorHists::Type::Muon, MuonsBX->size());
+  _Hists.FillNumber(L1ValidatorHists::Type::Egamma, EGammasBX->size());
+  _Hists.FillNumber(L1ValidatorHists::Type::Tau, TausBX->size());
+  _Hists.FillNumber(L1ValidatorHists::Type::Jet, JetsBX->size());
+
+  //For gen jet
+
+  for(uint i=0; i < GenJets->size(); i++){
```
Comment (davidlange6) [inline Validation/L1T/plugins/L1Validator.cc]: @ly615 - please use

for ( auto &Genjet: GenJets ) { 
}

(and line 122 can be removed)

File: Validation/L1T/plugins/L1Validator.cc [hunk 2]
```diff
@@ -82,78 +81,141 @@ void L1Validator::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetu
...
+     const reco::GenJet *Genjet = &GenJets->at(i);
+     if(fabs(Genjet->eta())>4.7) continue;
+
+     double minDR = 999.0;
+     const l1t::Jet        *L1Part=NULL;
+     for(int iBx = JetsBX->getFirstBX();  iBx<=JetsBX->getLastBX(); ++iBx){
+          if(iBx>0) continue;
+          for(std::vector<l1t::Jet>::const_iterator jet = JetsBX->begin(iBx); jet != JetsBX->end(iBx); ++jet){
```
Comment (davidlange6) [inline Validation/L1T/plugins/L1Validator.cc]: similarly here

for ( auto &jet : JetsBX)

File: Validation/L1T/plugins/L1Validator.cc [hunk 3]
```diff
@@ -82,78 +81,141 @@ void L1Validator::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetu
...
-      }
-    }else continue;
-
-    //cout << GenPart->pt() << '\t' << GenPart->eta() << '\t' << GenPart->phi() << '\t' << GenPart->pdgId() << endl;
+    double minDR = 999.0;
+    if(status==1 && abs(pdg)==13){  //Muon
+       if(fabs(GenPart->eta())>2.4) continue;
+       // calculate the isolation
+       double TotalIso30 = 0;
+       for(uint j=0; j<GenParticles->size(); j++){
```
Comment (davidlange6) [inline Validation/L1T/plugins/L1Validator.cc]: and here and line 163

File: Validation/L1T/plugins/L1Validator.cc [hunk 4]
```diff
@@ -82,78 +81,141 @@ void L1Validator::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetu
...
+			 minDR = idR;
+			 L1Part = &(*mu);
+		}
+	 		
+	  }
+          _Hists.Fill(L1ValidatorHists::Type::Muon, GenPart, L1Part);
+       }
+
+     }  else if(status==1 && (abs(pdg)==11 || pdg==22)){  //Egamma
+       if(fabs(GenPart->eta())>2.5) continue;
```
Comment (davidlange6) [inline Validation/L1T/plugins/L1Validator.cc]: please document these magic numbers

**Comment by davidlange6** [comment]
-1
pending answers to comments

File: Validation/L1T/plugins/L1Validator.cc (hunk 2 above)
Comment (rekovic) [inline Validation/L1T/plugins/L1Validator.cc]: @davidlange6  In this case, could we stay with the old style for loop, unless strong arguments against it ? 

I would agree that there are readability arguments for range-based for loops.
However, I have not seen it used with BXVector. ( JetBx is a BXVector pointer.)
BXVector has begin( ), end( ), begin(int), end(int) methods, but its use in range-based loops has not been tested, or I don't remember one.

File: Validation/L1T/plugins/L1Validator.cc (hunk 3 above)
Comment (rekovic) [inline Validation/L1T/plugins/L1Validator.cc]: range-base for loop OK here, but for line 163, same as comment above, could stay with old style for loop?

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
I guess we don't need to go with the same round of signatures since the last commit only 
addresses comments of @davidlange6, if David is OK with them.

**Comment by rekovic** [comment]
@davidlange6 We would like this for 9_1_0_pre3, if possible.

**Comment by civanch** [comment]
+1

**Comment by rekovic** [comment]
@perrotta, @dmitrijus 
Can you please sign again

**Comment by perrotta** [comment]
+1
(Changes applied since https://github.com/cms-sw/cmssw/pull/18156#issuecomment-292107318 arre fine with Reco)

**Comment by dmitrijus** [comment]
+1

**Comment by davidlange6** [comment]
+1
